File: clients/vscode/test/developerChecksClient.test.js
"use strict";
const { recipeExplanation, recipeExplanationText, recipeAssertionFindings } = require("../developerChecksClient");
const test = require("node:test");
const assert = require("node:assert/strict");
const { checkArguments, validateCheck, progressMessage, monitorCheck, provenanceSummary, comparisonSummary } = require("../developerChecksClient");
const session = "work-session-v2-" + "a".repeat(32);
const attempt = "check-" + "b".repeat(32);
test("recipe selection requires exact source and reference identities without execution consent", () => {
  const recipe = "saved-recipe:sha256:" + "e".repeat(64), image = "runtime-image:sha256:" + "f".repeat(64);
  const args = checkArguments(session, "prepare", image, { recipe, reference: attempt, absent: true });
  assert.deepEqual(args.slice(-5), ["--recipe", recipe, "--recipe-reference", attempt, "--absent"]);
  assert.ok(!args.includes("--confirm"));
  assert.ok(checkArguments(session, "prepare", image, { recipe, trace: false }).includes("--no-trace"));
  assert.deepEqual(checkArguments(session, "source", attempt, "lifecycle-e10:0").slice(-3), [attempt, "--source", "lifecycle-e10:0"]);
  assert.deepEqual(checkArguments(session, "source", attempt, "decision-q0:0").slice(-3), [attempt, "--source", "decision-q0:0"]);
  assert.throws(() => checkArguments(session, "source", attempt, "../secret"), /source/);
  assert.deepEqual(checkArguments(session, "recipes", { path: "groovy/postInit/Some File.groovy", reference: attempt }).slice(-3), ["--path=groovy/postInit/Some File.groovy", "--reference", attempt]);
  assert.throws(() => checkArguments(session, "prepare", image, { recipe, absent: true }), /retained/);
  assert.throws(() => checkArguments(session, "recipes", { path: "groovy/postInit/../../secret.groovy" }), /source path/);
  assert.throws(() => checkArguments(session, "prepare", image, { recipe: "latest" }), /exact saved recipe/);
});
test("comparison requires two exact runs without granting runtime consent", () => {
  const reference = "check-" + "d".repeat(32);
  assert.deepEqual(checkArguments(session, "compare", attempt, reference).slice(-4), ["compare", attempt, "--reference", reference]);
  assert.throws(() => checkArguments(session, "compare", attempt, attempt), /distinct/);
  assert.throws(() => checkArguments(session, "compare", attempt, "latest"), /exact/);
  assert.ok(!checkArguments(session, "compare", attempt, reference).includes("--confirm"));
  assert.equal(checkArguments(session, "history").at(-1), "history");
});
test("provenance and comparison summaries preserve uncertainty and original failures", () => {
  const value = { candidate: { source: { revision: "exact", dirty: true } }, provenance: {
    runtime: { pack: { name: "Supersymmetry", version: "0.1.16.15" }, platform: { id: "cleanroom", version: "0.6.12-alpha", java: { JAVA_RUNTIME_VERSION: "25.0.4+7" } } },
    image: { id: "exact-image" }, source_labels: { local_tags: ["latest"], release_verified: false },
  } };
  assert.match(provenanceSummary(value), /saved working-tree changes/);
  assert.match(provenanceSummary(value), /release\/latest not verified/);
  assert.match(provenanceSummary(value), /25\.0\.4\+7/);
  const comparison = { state: "partial", reference: { state: "inconclusive" }, candidate: { state: "failed" }, counts: { "newly-observed": 1 }, reasons: [] };
  assert.match(comparisonSummary(comparison), /inconclusive → candidate failed/);
  assert.match(comparisonSummary(comparison), /outcomes are unchanged/);
});
test("comparison response cannot promote outcomes or reuse historical result formats", () => {
  const value = { format: "workbench-developer-action-v1", exit_code: 0, context: { selection: { pack_uri: "file:///pack" } }, result: { format: "workbench-check-comparison-v2", authority: { runtime_launched: false, source_mutated: false, qualification_granted: false, outcomes_promoted: false } } };
  assert.equal(validateCheck(value, "file:///pack"), value.result);
  value.result.authority.outcomes_promoted = true;
  assert.throws(() => validateCheck(value, "file:///pack"), /authority/);
  value.result.format = "workbench-saved-check-result-v2";
  assert.throws(() => validateCheck(value, "file:///pack"), /historical/);
});
test("progress is tied to one request and arbitrary retained report paths are safe", () => {
  const result = { format: "workbench-check-live-status-v1", attempt_id: attempt, request_id: "request", state: "running", progress: { attempt_id: attempt, request_id: "request", observation: { format: "workbench-check-observation-v1", summary: "Compiling saved source" } } };
  assert.match(progressMessage(result, attempt, "request"), /Compiling saved source/);
  assert.throws(() => progressMessage(result, attempt, "other"), /another check/);
  assert.equal(checkArguments(session, "progress", attempt).at(-2), "progress");
  assert.equal(checkArguments(session, "log", attempt, "crash-reports/crash-client.txt").at(-1), "crash-reports/crash-client.txt");
  assert.throws(() => checkArguments(session, "log", attempt, "crash-reports/../secret"), /retained log/);
});
test("progress errors neither cancel execution nor outlive its result", async () => {
  let complete, polls = 0;
  const messages = [];
  const execution = new Promise((resolve) => { complete = resolve; });
  const running = monitorCheck(() => execution, async () => { polls += 1; throw new Error("temporarily unavailable"); }, (message) => { messages.push(message); complete("retained result"); }, attempt, "request", 1);
  assert.equal(await running, "retained result");
  assert.equal(polls, 1);
  assert.match(messages[0], /execution continues/);
  await new Promise((resolve) => setTimeout(resolve, 10));
  assert.equal(polls, 1);
});
test("environment preparation requires distinct exact consent and preserves path tokens", () => {
  const environment = "environment-" + "d".repeat(32);
  const request = "environment-request:sha256:" + "e".repeat(64);
  assert.throws(() => checkArguments(session, "prepare-environment", environment), /Confirm/);
  assert.throws(() => checkArguments(session, "prepare-environment", environment, "saved-check-request:sha256:" + "e".repeat(64)), /Confirm/);
  assert.deepEqual(checkArguments(session, "prepare-environment", environment, request).slice(-3), [environment, "--confirm", request]);
  const args = checkArguments(session, "plan-environment", { prism: "/tool path/PrismLauncher", packwiz: "/tools/packwiz", java: "/jdk/bin/java", accounts: "/private/accounts.json", seeds: ["/seed path/mods"] });
  assert.ok(args.includes("--prism=/tool path/PrismLauncher"));
  assert.ok(args.includes("--seed=/seed path/mods"));
  assert.deepEqual(checkArguments(session, "environment-cancel", environment).slice(-2), ["environment-cancel", environment]);
});
test("check execution requires exact explicit consent, cancellation is separate", () => {
  assert.throws(() => checkArguments(session, "execute", attempt), /consent/);
  const id = "saved-check-request:sha256:" + "c".repeat(64);
  assert.deepEqual(checkArguments(session, "execute", attempt, id), ["context", "run", session, "--", "checks", "execute", attempt, "--confirm", id]);
  assert.deepEqual(checkArguments(session, "cancel", attempt).slice(-3), ["checks", "cancel", attempt]);
  assert.throws(() => checkArguments(session, "show", "../escape"), /exact/);
});
test("check result rejects workspace retargeting and source mutation claims", () => {
  const value = { format: "workbench-developer-action-v1", exit_code: 0, context: { selection: { pack_uri: "file:///pack" } }, result: { format: "workbench-saved-check-result-v4", workspace_uri: "file:///pack", authority: { source_mutated: false, construction_authorized: false, qualification_granted: false } } };
  assert.equal(validateCheck(value, "file:///pack"), value.result);
  assert.throws(() => validateCheck(value, "file:///other"), /workspace/);
  value.result.authority.source_mutated = true;
  assert.throws(() => validateCheck(value, "file:///pack"), /authority/);
});
test("recipe explanations retain owner text and never promote inconclusive startup", () => {
  const report = { format: "workbench-check-explanation-v1", text: "Duration (ticks): expected 200; observed 240", limitations: ["Not source-causation proof."], sections: [{ id: "lookup-q0", text: "Observed winner r0", properties: [{ label: "Duration (ticks)", expected: "200", observed: "240" }] }] };
  const result = { candidate_id: "current", state: "inconclusive", assertions: { state: "mismatched", reasons: [], expectation: { source: { candidate_id: "current" }, subject: { location: { path: "Probe.groovy" } } }, observation: { details: { explanation: report } } } };
  assert.equal(recipeExplanation(result), report);
  assert.match(recipeExplanationText(result), /Startup: inconclusive/);
  assert.match(recipeExplanationText(result), /expected 200; observed 240/);
  assert.match(recipeExplanationText(result, report.sections[0]), /Not source-causation proof/);
  assert.equal(recipeAssertionFindings(result).length, 1);
  result.assertions.expectation.source.candidate_id = "retained-old-source";
  assert.deepEqual(recipeAssertionFindings(result), []);
  result.assertions.expectation.source.candidate_id = "current";
  result.assertions.state = "inconclusive";
  assert.deepEqual(recipeAssertionFindings(result), []);
  report.format = "future";
  assert.throws(() => recipeExplanation(result), /Unsupported/);
});

test("actual lookup explanations stay owner-produced and do not certify machine execution", () => {
  const text = "Original findRecipe invocations: 1; selected recipe: o0. Accepting recipe o1 was not evaluated. Not machine execution validation.";
  const section = { id: "decision-q0", text, properties: [], sources: [] };
  const report = { format: "workbench-check-explanation-v1", text, sections: [section], limitations: ["Scenario-bound lookup only."] };
  const result = { state: "inconclusive", assertions: { state: "matched", reasons: [], observation: { details: { explanation: report } } } };
  assert.equal(recipeExplanation(result), report);
  assert.ok(recipeExplanationText(result, section).includes(text));
  assert.match(recipeExplanationText(result, section), /Startup: inconclusive/);
  assert.equal(result.state, "inconclusive");
});

test("older assertions are not silently reinterpreted into new explanations", () => {
  const result = { state: "failed", assertions: { state: "inconclusive", reasons: [], observation: { details: {} } } };
  assert.equal(recipeExplanation(result), null);
  assert.match(recipeExplanationText(result), /fresh check is required/);
});

File: clients/intellij-community/src/test/java/dev/cleanroommc/workbench/intellij/community/DeveloperChecksClientTest.java
package dev.cleanroommc.workbench.intellij.community;

import org.junit.Test;
import static org.junit.Assert.*;

public class DeveloperChecksClientTest {
    @Test public void explanationsUseRetainedTextAndNeverAnnotateOldExpectedSource() {
        var result = com.google.gson.JsonParser.parseString("""
            {"candidate_id":"current","state":"inconclusive","assertions":{"state":"mismatched","reasons":[],
              "expectation":{"source":{"candidate_id":"current"},"subject":{"location":{"path":"Probe.groovy"}}},
              "observation":{"details":{"explanation":{"format":"workbench-check-explanation-v1",
                "text":"Duration (ticks): expected 200; observed 240","limitations":["Not source-causation proof."],
                "sections":[{"id":"lookup-q0","text":"Observed winner r0","properties":[{"label":"Duration (ticks)","expected":"200","observed":"240"}]}]}}}}}
            """).getAsJsonObject();
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, null).contains("Startup: inconclusive"));
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, null).contains("expected 200; observed 240"));
        var report = DeveloperChecksClient.recipeExplanation(result);
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, report.getAsJsonArray("sections").get(0).getAsJsonObject()).contains("Not source-causation proof"));
        assertEquals(1, DeveloperChecksClient.findings(result).size());
        result.getAsJsonObject("assertions").getAsJsonObject("expectation").getAsJsonObject("source").addProperty("candidate_id", "old");
        assertEquals(0, DeveloperChecksClient.findings(result).size());
        result.getAsJsonObject("assertions").getAsJsonObject("expectation").getAsJsonObject("source").addProperty("candidate_id", "current");
        result.getAsJsonObject("assertions").addProperty("state", "inconclusive");
        assertEquals(0, DeveloperChecksClient.findings(result).size());
        report.addProperty("format", "future");
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.recipeExplanation(result));
    }
    @Test public void olderAssertionsAreNeverReinterpretedIntoExplanations() {
        var result = com.google.gson.JsonParser.parseString("""
            {"state":"failed","assertions":{"state":"inconclusive","reasons":[],"observation":{"details":{}}}}
            """).getAsJsonObject();
        assertNull(DeveloperChecksClient.recipeExplanation(result));
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, null).contains("fresh check is required"));
    }
    @Test public void actualLookupDecisionTextDoesNotCertifyMachineExecution() {
        var result = com.google.gson.JsonParser.parseString("""
            {"state":"inconclusive","assertions":{"state":"matched","reasons":[],
              "observation":{"details":{"explanation":{"format":"workbench-check-explanation-v1",
                "text":"Original findRecipe invocations: 1; selected recipe: o0. Not machine execution validation.",
                "limitations":["Scenario-bound lookup only."],
                "sections":[{"id":"decision-q0","text":"Accepting recipe o1 was not evaluated.","properties":[],"sources":[]}]}}}}}
            """).getAsJsonObject();
        var report = DeveloperChecksClient.recipeExplanation(result);
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, null).contains("Not machine execution validation"));
        assertTrue(DeveloperChecksClient.recipeExplanationText(result, report.getAsJsonArray("sections").get(0).getAsJsonObject()).contains("o1 was not evaluated"));
        assertEquals("inconclusive", result.get("state").getAsString());
    }
    @Test public void recipeSelectionUsesExactIdentitiesAndSeparateConsent() {
        String session = "work-session-v2-" + "a".repeat(32), image = "runtime-image:sha256:" + "b".repeat(64);
        var options = new com.google.gson.JsonObject();
        options.addProperty("recipe", "saved-recipe:sha256:" + "c".repeat(64));
        options.addProperty("reference", "check-" + "d".repeat(32)); options.addProperty("absent", true);
        var args = DeveloperChecksClient.arguments(session, "prepare", image, null, options);
        assertTrue(args.contains("--recipe-reference")); assertTrue(args.contains("--absent")); assertFalse(args.contains("--confirm"));
        options.addProperty("trace", false);
        assertTrue(DeveloperChecksClient.arguments(session, "prepare", image, null, options).contains("--no-trace"));
        assertTrue(DeveloperChecksClient.arguments(session, "source", "check-" + "d".repeat(32), "lifecycle-e10:0").contains("--source"));
        assertTrue(DeveloperChecksClient.arguments(session, "source", "check-" + "d".repeat(32), "decision-q0:0").contains("--source"));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "source", "check-" + "d".repeat(32), "../secret"));
        options.remove("reference");
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "prepare", image, null, options));
        var source = new com.google.gson.JsonObject(); source.addProperty("path", "groovy/postInit/Some File.groovy");
        assertTrue(DeveloperChecksClient.recipeArguments(source, true).contains("--path=groovy/postInit/Some File.groovy"));
        source.addProperty("path", "groovy/postInit/../secret.groovy");
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.recipeArguments(source, true));
    }
    @Test public void comparisonRequiresDistinctRunsWithoutExecutionConsent() {
        String session = "work-session-v2-" + "a".repeat(32), candidate = "check-" + "b".repeat(32), reference = "check-" + "c".repeat(32);
        var arguments = DeveloperChecksClient.arguments(session, "compare", candidate, reference);
        assertEquals(reference, arguments.get(arguments.size() - 1));
        assertTrue(arguments.contains("--reference"));
        assertFalse(arguments.contains("--confirm"));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "compare", candidate, candidate));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "compare", candidate, "latest"));
        assertTrue(DeveloperChecksClient.arguments(session, "history", null, null).contains("history"));
    }
    @Test public void provenanceAndComparisonDoNotClaimLatestOrPromoteFailedRuns() {
        var value = com.google.gson.JsonParser.parseString("""
            {"candidate":{"source":{"revision":"exact","dirty":true}},"provenance":{
              "runtime":{"pack":{"name":"Supersymmetry","version":"0.1.16.15"},"platform":{"id":"cleanroom","version":"0.6.12-alpha","java":{"JAVA_RUNTIME_VERSION":"25.0.4+7"}}},
              "image":{"id":"image"},"source_labels":{"local_tags":["latest"],"release_verified":false}}}
            """).getAsJsonObject();
        assertTrue(DeveloperChecksClient.provenanceSummary(value).contains("saved working-tree changes"));
        assertTrue(DeveloperChecksClient.provenanceSummary(value).contains("release/latest not verified"));
        var comparison = com.google.gson.JsonParser.parseString("""
            {"state":"partial","reference":{"state":"inconclusive"},"candidate":{"state":"failed"},"counts":{"newly-observed":1},"reasons":[]}
            """).getAsJsonObject();
        assertTrue(DeveloperChecksClient.comparisonSummary(comparison).contains("candidate failed"));
        assertTrue(DeveloperChecksClient.comparisonSummary(comparison).contains("outcomes are unchanged"));
    }
    @Test public void progressAndReportSelectionStayBoundToExactRequest() {
        String session = "work-session-v2-" + "a".repeat(32), attempt = "check-" + "b".repeat(32);
        var result = com.google.gson.JsonParser.parseString("""
            {"format":"workbench-check-live-status-v1","attempt_id":"ATTEMPT","request_id":"request","state":"running",
             "progress":{"attempt_id":"ATTEMPT","request_id":"request","observation":{"format":"workbench-check-observation-v1","summary":"Compiling saved source"}}}
            """.replace("ATTEMPT", attempt)).getAsJsonObject();
        assertTrue(DeveloperChecksClient.progressMessage(result, attempt, "request").contains("Compiling saved source"));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.progressMessage(result, attempt, "other"));
        assertTrue(DeveloperChecksClient.arguments(session, "progress", attempt, null).contains("progress"));
        assertTrue(DeveloperChecksClient.arguments(session, "log", attempt, "crash-reports/crash-client.txt").contains("crash-reports/crash-client.txt"));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "log", attempt, "crash-reports/../secret"));
    }
    @Test public void environmentPreparationRequiresItsOwnExactConsent() {
        String session = "work-session-v2-" + "a".repeat(32), attempt = "environment-" + "b".repeat(32);
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "prepare-environment", attempt, null));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "prepare-environment", attempt, "saved-check-request:sha256:" + "c".repeat(64)));
        var argv = DeveloperChecksClient.arguments(session, "prepare-environment", attempt, "environment-request:sha256:" + "c".repeat(64));
        assertEquals("--confirm", argv.get(argv.size() - 2));
        assertFalse(DeveloperChecksClient.arguments(session, "environment-cancel", attempt, null).contains("--confirm"));
    }
    @Test public void executionRequiresExactConsent() {
        String session = "work-session-v2-" + "a".repeat(32), attempt = "check-" + "b".repeat(32);
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments(session, "execute", attempt, null));
        var argv = DeveloperChecksClient.arguments(session, "execute", attempt, "saved-check-request:sha256:" + "c".repeat(64));
        assertEquals("--confirm", argv.get(argv.size() - 2));
        assertFalse(DeveloperChecksClient.arguments(session, "cancel", attempt, null).contains("--confirm"));
    }
    @Test public void neverAcceptsAmbiguousOrPathAttemptSelectors() {
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments("latest", "show", "../outside", null));
        assertThrows(IllegalArgumentException.class, () -> DeveloperChecksClient.arguments("work-session-v2-" + "a".repeat(32), "show", "../outside", null));
    }
}
